ipld/read_test.go

```diff
@@ -275,13 +275,145 @@ func TestGetLeavesByNamespace(t *testing.T) {
 				for _, node := range nodes {
 					// TODO @renaynay: nID is prepended twice for some reason.
 					share := node.RawData()[1:]
-					assert.Equal(t, expected, share[8:])
+					assert.Equal(t, expected, share[NamespaceSize:])
 				}
 			}
 		})
 	}
 }
 
+func TestGetLeavesByNamespace_AbsentNamespaceId(t *testing.T) {
+
+	t.Run("Namespace id less than the minimum namespace in data", func(t *testing.T) {
+
+		rawData := RandNamespacedShares(t, 16).Raw()
+
+		// replace minimum NID with the second minimum NID
+		minNid := make([]byte, NamespaceSize)
+		copy(minNid, rawData[0][:NamespaceSize])
+		secondMinNid := rawData[1][:NamespaceSize]
+		copy(rawData[0][:NamespaceSize], secondMinNid[:NamespaceSize])
+
+		dag, dah := putErasuredDataToDag(t, rawData)
+
+		assertNoRowContainsNID(t, dag, dah, minNid)
+	})
+
+	t.Run("Namespace id greater than the maximum namespace in data", func(t *testing.T) {
+
+		rawData := RandNamespacedShares(t, 16).Raw()
+
+		// replace maximum NID with the second maximum NID
+		lastItemIndex := len(rawData) - 1
+		maxNid := make([]byte, NamespaceSize)
+		copy(maxNid, rawData[lastItemIndex][:NamespaceSize])
+		secondMaxNid := rawData[lastItemIndex-1][:NamespaceSize]
+		copy(rawData[lastItemIndex][:NamespaceSize], secondMaxNid[:NamespaceSize])
+
+		dag, dah := putErasuredDataToDag(t, rawData)
+
+		assertNoRowContainsNID(t, dag, dah, maxNid)
+	})
+
+	t.Run("Namespace id in range but still missing", func(t *testing.T) {
+
+		rawData := RandNamespacedShares(t, 16).Raw()
+
+		// replace every appearance of someNid with the nextNid so that someNid, which we know
+		// that will be in range, will not exist in the namespace range
+		someNid := make([]byte, NamespaceSize)
+		copy(someNid, rawData[len(rawData)/2][:NamespaceSize])
+		nextNid := rawData[(len(rawData)/2)+1][:NamespaceSize]
+		for _, nspace := range rawData {
+			if bytes.Equal(nspace[:NamespaceSize], someNid) {
+				copy(nspace[:NamespaceSize], nextNid)
+			}
+		}
+
+		dag, dah := putErasuredDataToDag(t, rawData)
+
+		assertNoRowContainsNID(t, dag, dah, someNid)
+	})
```

Comment:
At least this test looks like a natural candidate for [table driven tests](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests).

Reply:
Sounds good. Converting to table driven tests. good to know!

Reply:
Done.